engines/ags/lib/std/set.h

```diff
@@ -55,11 +60,18 @@ class set : public Common::SortedArray<T, const T &> {
 	 * Locate an item in the set
 	 */
 	iterator find(const T &item) {
-		iterator it;
-		for (it = this->begin(); it != this->end() && *it != item; ++it) {
+		iterator begin = this->begin();
+		iterator end = this->end();
+		while (begin < end) {
+			iterator mid = begin + (Common::distance(begin, end) / 2);
+			if (ComparatorFn(item, *mid))
+				end = mid;
+			else if (ComparatorFn(*mid, item))
+			 	begin = mid + 1;
+			else
+				return mid;
 		}
-
-		return it;
+		return this->end();
 	}
```

Comment:
I will need to double check this change.

I think it is fine, but using a binary search as above could change the result when an element is present multiple times in the container as the previous code always returned the first one, which may no longer be the case. However since `std::set` only has unique items, this should be fine for this specific container.

But where I am getting a bit confused is that we are using `Common::SortedArray` here to implement `std::set`, and I think it behaves like a `std::multiset` and can have the same item more than once? I am actually wondering if our implementation of `std::set` here could be faulty and cause issues in AGS games. This is not an issue introduced in this pull request, but one that the change here is highlighting.

Reply:
I agree, the implementation does allow more items that are equal to each other to be inserted. I also had to replace `std::multiset`, and for now a simple typedef to our set implementation worked. One problem I thought about is that iterating through the container could cause items to be processed multiple times when no more than one is expected.